plugins/scaffolder-backend/src/scaffolder/actions/builtin/catalog/write.ts

```diff
@@ -19,32 +19,48 @@ import { createTemplateAction } from '../../createTemplateAction';
 import * as yaml from 'yaml';
 import { Entity } from '@backstage/catalog-model';
 import { resolveSafeChildPath } from '@backstage/backend-common';
+import { z } from 'zod';
+import zodToJsonSchema from 'zod-to-json-schema';
 
 /**
  * Writes a catalog descriptor file containing the provided entity to a path in the workspace.
  * @public
  */
+
+// This function will need its own file to be instantiated in other scafolder action files in the future
+// Note: Can be used for both input and output keys, example below (lines 42-45 and 52).
+function convertZodToJson(zodSchema: z.ZodObject<{}>) {
```

Comment:
I'd recommend taking a look at how the frontend does this for `CustomFieldExtension` courtesy of @kuangp.

https://github.com/backstage/backstage/blob/da25293910f208d00296af7c7de497a8f8ff7278/plugins/scaffolder/src/components/fields/utils.ts#L35-L62

It might be worth then digging into how that type is used:

https://github.com/backstage/backstage/blob/da25293910f208d00296af7c7de497a8f8ff7278/plugins/scaffolder/src/components/fields/EntityTagsPicker/schema.ts#L16-L48

And maybe copying some of the implementation from there. This method and type helper could potentially live in `scaffolder-common` maybe as it's something that we're going to have for both the `frontend` and `backend` of the scaffolder plugin.